extensions/ext/GLSL_EXT_spirv_intrinsics.txt

```diff
@@ -0,0 +1,1020 @@
+
+            set = literal-string
+            id = integer-constant-expression
+
+        The `set` qualifier indicates the name of an extended
+        instruction set that the type's instruction belongs to.
+        If the `set` qualifier is omitted, or the string is empty,
+        no extended instruction set will be imported, and the instruction will
+        be selected from the core instruction set.
+
+        The `id` is the id of the type instruction in the selected
+        instruction set.
+
+        Parameters to this qualifier must be either literal values, or constant
+        expressions as defined in section 4.3.3, Constant Expressions.
+
+        When generating SPIR-V, the compiler will generate type
+        instructions for these types in the following manner:
+
+          * If any variable is declared with a SPIR-V type that requires an
+            extended instruction set (declared with the `set` qualifier),
+            it will add an *OpExtInstImport* instruction with the same
+            literal string.
+          * For variables declared with a SPIR-V type that requires an
+            extended instruction set, an *OpExtInst* instruction will be
+            inserted into the resulting SPIR-V, declaring the type.
+          * For variables declared with a SPIR-V type that is defined by
+            instructions in the core instruction set, an instruction with
+            the declared `id` will be inserted into the resulting SPIR-V.
+
+        Any extensions or capabilities needed by the type must be declared
+        using SPIR-V Requirements Declarations (see section 3.9).
+
+        Types declared in this way do not implicitly allow any usage other
+        than variable declaration and as function parameters, and cannot be
+        l-values.
+        When used as parameters in user-defined functions, spir-v will
+        always use pass-by-reference semantics (as if all the declaration
+        used the `spirv_by_reference` qualifier).
+        The `spirv_by_reference` qualifier must: still be present on
+        functions defined by the `spirv_instruction` qualifier that require
+        it.
+
+Additions to Chapter 6 of the OpenGL Shading Language Specification
+(Statements and Structure)
+
+    Modify section 6.1.1, Function Calling Conventions
+    
+        In the definition of `parameter-qualifier`, add the following:
+        
+            spirv_by_reference spirv_literal
+            
+        At the end of this section, add the following text:
+
+            When generating SPIR-V, the `spirv_by_reference` qualifier indicates
+            to the compiler that the variable should be passed as a SPIR-V
+            pointer rather than generating *OpLoad*/*OpStore* instructions
+            for the underlying object.
+            
+            When generating SPIR-V, the `spirv_literal` qualifier indicates
+            to the compiler that the variable should be compiled to a literal
```

Comment:
I find this a bit confusing. This applies to function parameters, so it should be saying that values passed to the parameter (rather than "the variable") will be compiled to a literal? I don't think SPIR-V has a way of accepting literals as function arguments, rather than the IDs of values, does it? If not then I'm not sure what this is really for.

Reply:
> I find this a bit confusing. This applies to function parameters, so it should be saying that values passed to the parameter (rather than "the variable") will be compiled to a literal?

I suppose that's more accurate so sure.

> I don't think SPIR-V has a way of accepting literals as function arguments, rather than the IDs of values, does it?

User defined functions, no, but built-in instructions yes. An inlining pass could theoretically make it relevant on a function declaration, though that's purely hypothetical atm. Maybe I could caveat this?

Reply:
Oh wait I already did caveat this - it's only allowed on spirv instructions 😅